// mc-build/dist/haxpression/utils/Chars.js
import {Register} from "../../genes/Register.js"

const $global = Register.$global

export const Chars = Register.global("$hxClasses")["haxpression.utils.Chars"] = 
class Chars {
	static isDecimalDigit(charCode) {
		if (charCode >= 48) {
			return charCode <= 57;
		} else {
			return false;
		};
	}
	static isUpperCaseLetter(charCode) {
		if (charCode >= 65) {
			return charCode <= 90;
		} else {
			return false;
		};
	}
	static isLowerCaseLetter(charCode) {
		if (charCode >= 97) {
			return charCode <= 122;
		} else {
			return false;
		};
	}
	static isIdentifierStart(charCode) {
		if (!(charCode == Chars.DOLLAR_CODE || charCode == Chars.UNDERSCORE_CODE || Chars.isUpperCaseLetter(charCode) || Chars.isLowerCaseLetter(charCode))) {
			return Chars.OTHER_IDENTIFIER_START_CODES.includes(charCode);
		} else {
			return true;
		};
	}
	static isIdentifierPart(charCode) {
		if (!(Chars.isIdentifierStart(charCode) || Chars.isDecimalDigit(charCode))) {
			return Chars.OTHER_IDENTIFIER_PART_CODES.includes(charCode);
		} else {
			return true;
		};
	}
	static isWhiteSpace(charCode) {
		if (!(charCode == Chars.TAB_CODE || charCode == Chars.LF_CODE || charCode == Chars.CR_CODE)) {
			return charCode == Chars.SPACE_CODE;
		} else {
			return true;
		};
	}
	static get __name__() {
		return "haxpression.utils.Chars"
	}
	get __class__() {
		return Chars
	}
}


Chars.TAB_CODE = 9
Chars.LF_CODE = 10
Chars.CR_CODE = 13
Chars.SPACE_CODE = 32
Chars.EXCLAMATION_POINT_CODE = 33
Chars.DOUBLE_QUOTE_CODE = 34
Chars.DOLLAR_CODE = 36
Chars.SINGLE_QUOTE_CODE = 39
Chars.OPEN_PAREN_CODE = 40
Chars.CLOSE_PAREN_CODE = 41
Chars.COMMA_CODE = 44
Chars.PERIOD_CODE = 46
Chars.COLON_CODE = 58
Chars.SEMICOLON_CODE = 59
Chars.QUESTION_MARK_CODE = 63
Chars.OPEN_BRACKET_CODE = 91
Chars.CLOSE_BRACKET_CODE = 93
Chars.UNDERSCORE_CODE = 95
Chars.OTHER_IDENTIFIER_START_CODES = []
Chars.OTHER_IDENTIFIER_PART_CODES = [Chars.EXCLAMATION_POINT_CODE, Chars.DOLLAR_CODE, Chars.COLON_CODE]
