// mc-build/dist/EReg.js
import {Exception} from "./haxe/Exception.js"
import {Register} from "./genes/Register.js"
import {HxOverrides} from "./HxOverrides.js"

const $global = Register.$global

/**
The EReg class represents regular expressions.

While basic usage and patterns consistently work across platforms, some more
complex operations may yield different results. This is a necessary trade-
off to retain a certain level of performance.

EReg instances can be created by calling the constructor, or with the
special syntax `~/pattern/modifier`

EReg instances maintain an internal state, which is affected by several of
its methods.

A detailed explanation of the supported operations is available at
<https://haxe.org/manual/std-regex.html>
*/
export const EReg = Register.global("$hxClasses")["EReg"] = 
class EReg extends Register.inherits() {
	new(r, opt) {
		this.r = new RegExp(r, opt.split("u").join(""));
	}
	
	/**
	Tells if `this` regular expression matches String `s`.
	
	This method modifies the internal state.
	
	If `s` is `null`, the result is unspecified.
	*/
	match(s) {
		if (this.r.global) {
			this.r.lastIndex = 0;
		};
		this.r.m = this.r.exec(s);
		this.r.s = s;
		return this.r.m != null;
	}
	
	/**
	Returns the matched sub-group `n` of `this` EReg.
	
	This method should only be called after `this.match` or
	`this.matchSub`, and then operates on the String of that operation.
	
	The index `n` corresponds to the n-th set of parentheses in the pattern
	of `this` EReg. If no such sub-group exists, the result is unspecified.
	
	If `n` equals 0, the whole matched substring is returned.
	*/
	matched(n) {
		if (this.r.m != null && n >= 0 && n < this.r.m.length) {
			return this.r.m[n];
		} else {
			throw Exception.thrown("EReg::matched");
		};
	}
	
	/**
	Returns the part to the right of the last matched substring.
	
	If the most recent call to `this.match` or `this.matchSub` did not
	match anything, the result is unspecified.
	
	If the global g modifier was in place for the matching, only the
	substring to the right of the leftmost match is returned.
	
	The result does not include the matched part.
	*/
	matchedRight() {
		if (this.r.m == null) {
			throw Exception.thrown("No string matched");
		};
		let sz = this.r.m.index + this.r.m[0].length;
		return HxOverrides.substr(this.r.s, sz, this.r.s.length - sz);
	}
	
	/**
	Returns the position and length of the last matched substring, within
	the String which was last used as argument to `this.match` or
	`this.matchSub`.
	
	If the most recent call to `this.match` or `this.matchSub` did not
	match anything, the result is unspecified.
	
	If the global g modifier was in place for the matching, the position and
	length of the leftmost substring is returned.
	*/
	matchedPos() {
		if (this.r.m == null) {
			throw Exception.thrown("No string matched");
		};
		return {"pos": this.r.m.index, "len": this.r.m[0].length};
	}
	static get __name__() {
		return "EReg"
	}
	get __class__() {
		return EReg
	}
}
