// mc-build/dist/Reflect.js
import {Register} from "./genes/Register.js"

const $global = Register.$global

/**
The Reflect API is a way to manipulate values dynamically through an
abstract interface in an untyped manner. Use with care.

@see https://haxe.org/manual/std-reflection.html
*/
export const Reflect = Register.global("$hxClasses")["Reflect"] = 
class Reflect {
	
	/**
	Returns the value of the field named `field` on object `o`.
	
	If `o` is not an object or has no field named `field`, the result is
	null.
	
	If the field is defined as a property, its accessors are ignored. Refer
	to `Reflect.getProperty` for a function supporting property accessors.
	
	If `field` is null, the result is unspecified.
	*/
	static field(o, field) {
		try {
			return o[field];
		}catch (_g) {
			return null;
		};
	}
	
	/**
	Returns the value of the field named `field` on object `o`, taking
	property getter functions into account.
	
	If the field is not a property, this function behaves like
	`Reflect.field`, but might be slower.
	
	If `o` or `field` are null, the result is unspecified.
	*/
	static getProperty(o, field) {
		let tmp;
		if (o == null) {
			return null;
		} else {
			let tmp1;
			if (o.__properties__) {
				tmp = o.__properties__["get_" + field];
				tmp1 = tmp;
			} else {
				tmp1 = false;
			};
			if (tmp1) {
				return o[tmp]();
			} else {
				return o[field];
			};
		};
	}
	
	/**
	Returns the fields of structure `o`.
	
	This method is only guaranteed to work on anonymous structures. Refer to
	`Type.getInstanceFields` for a function supporting class instances.
	
	If `o` is null, the result is unspecified.
	*/
	static fields(o) {
		let a = [];
		if (o != null) {
			let hasOwnProperty = Object.prototype.hasOwnProperty;
			for( var f in o ) {;
			if (f != "__id__" && f != "hx__closures__" && hasOwnProperty.call(o, f)) {
				a.push(f);
			};
			};
		};
		return a;
	}
	
	/**
	Returns true if `f` is a function, false otherwise.
	
	If `f` is null, the result is false.
	*/
	static isFunction(f) {
		if (typeof(f) == "function") {
			return !(f.__name__ || f.__ename__);
		} else {
			return false;
		};
	}
	
	/**
	Tells if `v` is an object.
	
	The result is true if `v` is one of the following:
	
	- class instance
	- structure
	- `Class<T>`
	- `Enum<T>`
	
	Otherwise, including if `v` is null, the result is false.
	*/
	static isObject(v) {
		if (v == null) {
			return false;
		};
		let t = typeof(v);
		if (!(t == "string" || t == "object" && v.__enum__ == null)) {
			if (t == "function") {
				return (v.__name__ || v.__ename__) != null;
			} else {
				return false;
			};
		} else {
			return true;
		};
	}
	static get __name__() {
		return "Reflect"
	}
	get __class__() {
		return Reflect
	}
}
