// mc-build/dist/mcl/Globals.js
import {ArrayIterator} from "../haxe/iterators/ArrayIterator.js"
import {StringMap} from "../haxe/ds/StringMap.js"
import {ObjectMap} from "../haxe/ds/ObjectMap.js"
import {Exception} from "../haxe/Exception.js"
import {Register} from "../genes/Register.js"
import {ValueType, Type} from "../Type.js"
import {Std} from "../Std.js"

const $global = Register.$global

export const McIntIterator = Register.global("$hxClasses")["mcl.McIntIterator"] = 
class McIntIterator extends Register.inherits() {
	new(min, max) {
		this.min = min;
		this.max = max;
		this.current = min;
		this.offset = (min < max) ? 1 : -1;
	}
	hasNext() {
		if (this.offset == 1) {
			return this.current <= this.max;
		} else {
			return this.current >= this.max;
		};
	}
	next() {
		let result = this.current;
		if (!this.hasNext()) {
			throw Exception.thrown("No such element");
		};
		this.current += this.offset;
		return result;
	}
	static get __name__() {
		return "mcl.McIntIterator"
	}
	get __class__() {
		return McIntIterator
	}
}


export const McFloatIterator = Register.global("$hxClasses")["mcl.McFloatIterator"] = 
class McFloatIterator extends Register.inherits() {
	new(min, max, step) {
		this.min = min;
		this.max = max;
		this.current = min;
		this.offset = step;
		if (step < 0 && min < max) {
			throw Exception.thrown("Invalid step for range");
		};
		if (step > 0 && min > max) {
			throw Exception.thrown("Invalid step for range");
		};
	}
	hasNext() {
		if (this.offset > 0) {
			return this.current <= this.max;
		} else {
			return this.current >= this.max;
		};
	}
	next() {
		let result = this.current;
		if (!this.hasNext()) {
			throw Exception.thrown("No such element");
		};
		this.current += this.offset;
		return result;
	}
	static get __name__() {
		return "mcl.McFloatIterator"
	}
	get __class__() {
		return McFloatIterator
	}
}


export const Globals = Register.global("$hxClasses")["mcl.Globals"] = 
class Globals {
	static set(name, value) {
		Globals.map.inst.set(name, value);
	}
	static get(name) {
		return Globals.map.inst.get(name);
	}
	static has(name) {
		return Globals.map.inst.has(name);
	}
	static delete(name) {
		return Globals.map.inst["delete"](name);
	}
	static get __name__() {
		return "mcl.Globals"
	}
	get __class__() {
		return Globals
	}
}


Globals.loopVariants = (function($this) {var $r0
	let _g = new ObjectMap();
	_g.inst.set([ValueType.TInt, ValueType.TInt], function (args) {
		let min = args[0];
		let max = args[1];
		return new McIntIterator(min, max);
	});
	_g.inst.set([ValueType.TFloat, ValueType.TFloat, ValueType.TFloat], function (args) {
		let min = args[0];
		let max = args[1];
		let step = args[2];
		return new McFloatIterator(min, max, step);
	});
	_g.inst.set([ValueType.TClass(Array)], function (args) {
		return new ArrayIterator(args[0]);
	});
	_g.inst.set([ValueType.TObject], function (args) {
		let iterator = new ArrayIterator(Object.entries(args[0]));
		return iterator;
	});
	_g.inst.set([ValueType.TFunction], function (args) {
		let iterator = args[0];
		return new ArrayIterator(Array.from(iterator()));
	});
	
	$r0 = _g
	return $r0})(this)
Globals.map = (function($this) {var $r0
	let _g = new StringMap();
	_g.inst.set("REPEAT", function (...args) {
		let argCount = args.length;
		let map = Globals.loopVariants;
		let _g_map = map;
		let _g_keys = map.keys();
		while (_g_keys.hasNext()) {
			let key = _g_keys.next();
			let _g_value = _g_map.get(key);
			let _g_key = key;
			let overlod = _g_key;
			let handler = _g_value;
			if (overlod.length == argCount) {
				let failure = false;
				let _g = 0;
				let _g1 = argCount;
				while (_g < _g1) {
					let i = _g++;
					let t = Type["typeof"](args[i]);
					if (!(Type.enumEq(overlod[i], t) || overlod[i] == ValueType.TFloat && t == ValueType.TInt)) {
						failure = true;
						break;
					};
				};
				if (!failure) {
					return handler(args.slice());
				};
			};
		};
		let _this = args.slice();
		let result = new Array(_this.length);
		let _g = 0;
		let _g1 = _this.length;
		while (_g < _g1) {
			let i = _g++;
			result[i] = Std.string(Type["typeof"](_this[i]));
		};
		throw Exception.thrown("Invalid arguments for REPEAT (" + result.join(", ") + ")");
	});
	
	$r0 = _g
	return $r0})(this)
