// mc-build/dist/mcl/LibStore.js
import {FileSystem} from "../sys/FileSystem.js"
import {LibraryError} from "./error/LibraryError.js"
import {Tokenizer} from "./Tokenizer.js"
import {Parser} from "./Parser.js"
import {ErrorUtil, Compiler, McFile} from "./Compiler.js"
import {Path} from "../haxe/io/Path.js"
import {StringMap} from "../haxe/ds/StringMap.js"
import {Register} from "../genes/Register.js"
import * as Fs from "fs"
import {HxOverrides} from "../HxOverrides.js"

const $global = Register.$global

export const LibStore = Register.global("$hxClasses")["mcl.LibStore"] = 
class LibStore extends Register.inherits() {
	new(dir) {
		this.loadedLibs = new StringMap();
		this.libDir = dir;
	}
	lookup(id, pos, compiler) {
		if (this.loadedLibs.inst.has(id)) {
			return this.loadedLibs.inst.get(id).inst.get("mcblib/" + id + ".mcbt");
		};
		let p = Path.join([this.libDir, id]);
		if (FileSystem.exists(p)) {
			return this.loadLib(id, p, compiler, pos);
		};
		throw new LibraryError(ErrorUtil.format("Library not found: " + id, pos));
	}
	getFilesInDirectory(dir) {
		let files = Fs.readdirSync(dir);
		let result = [];
		let _g = 0;
		while (_g < files.length) {
			let f = files[_g];
			++_g;
			let p = Path.join([dir, f]);
			if (FileSystem.isDirectory(p)) {
				result = result.concat(this.getFilesInDirectory(p));
			} else {
				result.push(p);
			};
		};
		return result;
	}
	loadLib(id, p, compiler, pos) {
		let baseDir = Path.join([p, "src", "mcblib"]);
		let srcDir = Path.join([p, "src"]);
		if (!FileSystem.exists(baseDir)) {
			throw new LibraryError("Library " + id + " does not have a src/mcblib folder");
		};
		let c = new Compiler(p, {});
		let files = this.getFilesInDirectory(baseDir);
		let result = new StringMap();
		let _g = 0;
		while (_g < files.length) {
			let f = files[_g];
			++_g;
			let ext = Path.extension(f);
			if (ext != "mbt" && ext != "mcbt") {
				continue;
			};
			let tokens = Tokenizer.tokenize(Fs.readFileSync(f, {"encoding": "utf8"}), f);
			let ast = (ext == "mcb") ? Parser.parseMcbFile(tokens) : Parser.parseMcbtFile(tokens);
			let mcFile = new McFile(f, ast);
			mcFile.setup(compiler);
			let finalPath = HxOverrides.substr(f, srcDir.length + 1, null);
			mcFile.name = finalPath;
			result.inst.set(finalPath, mcFile);
		};
		this.loadedLibs.inst.set(id, result);
		return result.inst.get("mcblib/" + id + ".mcbt");
	}
	static get __name__() {
		return "mcl.LibStore"
	}
	get __class__() {
		return LibStore
	}
}
