// mc-build/dist/mcl/TagManager.js
import {CompilerError} from "./error/CompilerError.js"
import {ErrorUtil} from "./Compiler.js"
import {Path} from "../haxe/io/Path.js"
import {StringMap} from "../haxe/ds/StringMap.js"
import {Register} from "../genes/Register.js"

const $global = Register.$global

export const TagManager = Register.global("$hxClasses")["mcl.TagManager"] = 
class TagManager extends Register.inherits() {
	new() {
		this.tagEntries = new StringMap();
	}
	ensureTag(tag, context) {
		let colonIndex = tag.indexOf(":");
		if (colonIndex == -1) {
			tag = context.namespace + ":" + context.path.concat([tag]).join("/");
		} else if (colonIndex != tag.lastIndexOf(":")) {
			let pos = null;
			throw new CompilerError(ErrorUtil.formatContext("Invalid tag name: " + tag, pos, context), false, [pos].concat(context.stack));
		};
		if (!this.tagEntries.inst.has(tag)) {
			this.tagEntries.inst.set(tag, {"entries": new Set(), "replace": false});
		};
		return tag;
	}
	addTagEntry(tag, entry, context, replace) {
		if (replace == null) {
			replace = false;
		};
		tag = this.ensureTag(tag, context);
		this.tagEntries.inst.get(tag).entries.add({"value": entry, "replace": replace});
	}
	setTagReplace(tag, context, replace) {
		tag = this.ensureTag(tag, context);
		this.tagEntries.inst.get(tag).replace = replace;
	}
	writeTagFiles(compiler) {
		let map = this.tagEntries;
		let _g_map = map;
		let _g_keys = map.keys();
		while (_g_keys.hasNext()) {
			let key = _g_keys.next();
			let _g_value = _g_map.get(key);
			let _g_key = key;
			let k = _g_key;
			let v = _g_value;
			let segments = k.split(":");
			let tmp = segments.length != 2;
			let namespace = segments[0];
			let tag = segments[1];
			let tagPath = Path.join(["data", namespace, "tags", (compiler.config.features.useFolderRenames48) ? "function" : "functions", tag + ".json"]);
			let compiler1 = compiler.io;
			let _g = [];
			let jsIterator = v.entries.values();
			let _g_jsIterator = jsIterator;
			let _g_lastStep = jsIterator.next();
			while (!_g_lastStep.done) {
				let v = _g_lastStep.value;
				_g_lastStep = _g_jsIterator.next();
				let entry = v;
				let entry1 = entry.replace;
				_g.push(entry.value);
			};
			compiler1.write(tagPath, JSON.stringify({"values": _g, "replace": (v.replace) ? true : undefined}));
		};
	}
	static get __name__() {
		return "mcl.TagManager"
	}
	get __class__() {
		return TagManager
	}
}
