// mc-build/dist/sys/FileSystem.js
import * as Path from "path"
import {Exception} from "../haxe/Exception.js"
import {Register} from "../genes/Register.js"
import * as Fs from "fs"

const $global = Register.$global

/**
This class provides information about files and directories.

If `null` is passed as a file path to any function in this class, the
result is unspecified, and may differ from target to target.

See `sys.io.File` for the complementary file API.
*/
export const FileSystem = Register.global("$hxClasses")["sys.FileSystem"] = 
class FileSystem {
	
	/**
	Returns `true` if the file or directory specified by `path` exists.
	*/
	static exists(path) {
		try {
			Fs.accessSync(path);
			return true;
		}catch (_g) {
			return false;
		};
	}
	
	/**
	Returns `true` if the file or directory specified by `path` is a directory.
	
	If `path` is not a valid file system entry or if its destination is not
	accessible, an exception is thrown.
	*/
	static isDirectory(path) {
		try {
			return Fs.statSync(path).isDirectory();
		}catch (_g) {
			return false;
		};
	}
	
	/**
	Creates a directory specified by `path`.
	
	This method is recursive: The parent directories don't have to exist.
	
	If the directory cannot be created, an exception is thrown.
	*/
	static createDirectory(path) {
		try {
			Fs.mkdirSync(path);
		}catch (_g) {
			let e = Exception.caught(_g).unwrap();
			if (e.code == "ENOENT") {
				FileSystem.createDirectory(Path.dirname(path));
				Fs.mkdirSync(path);
			} else {
				let stat;
				try {
					stat = Fs.statSync(path);
				}catch (_g) {
					throw e;
				};
				if (!stat.isDirectory()) {
					throw e;
				};
			};
		};
	}
	
	/**
	Deletes the directory specified by `path`. Only empty directories can
	be deleted.
	
	If `path` does not denote a valid directory, or if that directory cannot
	be deleted, an exception is thrown.
	*/
	static deleteDirectory(path) {
		if (FileSystem.exists(path)) {
			let _g = 0;
			let _g1 = Fs.readdirSync(path);
			while (_g < _g1.length) {
				let file = _g1[_g];
				++_g;
				let curPath = path + "/" + file;
				if (FileSystem.isDirectory(curPath)) {
					FileSystem.deleteDirectory(curPath);
				} else {
					Fs.unlinkSync(curPath);
				};
			};
			Fs.rmdirSync(path);
		};
	}
	static get __name__() {
		return "sys.FileSystem"
	}
	get __class__() {
		return FileSystem
	}
}
