// mc-build/dist/haxe/ds/List.js
import {Register} from "../../genes/Register.js"
import {Std} from "../../Std.js"

const $global = Register.$global

/**
A linked-list of elements. The list is composed of element container objects
that are chained together. It is optimized so that adding or removing an
element does not imply copying the whole list content every time.

@see https://haxe.org/manual/std-List.html
*/
export const List = Register.global("$hxClasses")["haxe.ds.List"] = 
class List extends Register.inherits() {
	new() {
		this.length = 0;
	}
	
	/**
	Adds element `item` at the end of `this` List.
	
	`this.length` increases by 1.
	*/
	add(item) {
		let x = new ListNode(item, null);
		if (this.h == null) {
			this.h = x;
		} else {
			this.q.next = x;
		};
		this.q = x;
		this.length++;
	}
	
	/**
	Adds element `item` at the beginning of `this` List.
	
	`this.length` increases by 1.
	*/
	push(item) {
		let x = new ListNode(item, this.h);
		this.h = x;
		if (this.q == null) {
			this.q = x;
		};
		this.length++;
	}
	
	/**
	Returns the first element of `this` List, or null if no elements exist.
	
	This function does not modify `this` List.
	*/
	first() {
		if (this.h == null) {
			return null;
		} else {
			return this.h.item;
		};
	}
	
	/**
	Returns the first element of `this` List, or null if no elements exist.
	
	The element is removed from `this` List.
	*/
	pop() {
		if (this.h == null) {
			return null;
		};
		let x = this.h.item;
		this.h = this.h.next;
		if (this.h == null) {
			this.q = null;
		};
		this.length--;
		return x;
	}
	
	/**
	Tells if `this` List is empty.
	*/
	isEmpty() {
		return this.h == null;
	}
	
	/**
	Returns a string representation of `this` List.
	
	The result is enclosed in { } with the individual elements being
	separated by a comma.
	*/
	toString() {
		let s_b = "";
		let first = true;
		let l = this.h;
		s_b += "{";
		while (l != null) {
			if (first) {
				first = false;
			} else {
				s_b += ", ";
			};
			s_b += Std.string(Std.string(l.item));
			l = l.next;
		};
		s_b += "}";
		return s_b;
	}
	static get __name__() {
		return "haxe.ds.List"
	}
	get __class__() {
		return List
	}
}


export const ListNode = Register.global("$hxClasses")["haxe.ds._List.ListNode"] = 
class ListNode extends Register.inherits() {
	new(item, next) {
		this.item = item;
		this.next = next;
	}
	static get __name__() {
		return "haxe.ds._List.ListNode"
	}
	get __class__() {
		return ListNode
	}
}
