// mc-build/dist/mcl/args/TemplateArgument.js
import {LiteralTemplateArgument} from "./LiteralTemplateArgument.js"
import {StringMap} from "../../haxe/ds/StringMap.js"
import {Exception} from "../../haxe/Exception.js"
import {Register} from "../../genes/Register.js"
import {Type} from "../../Type.js"

const $global = Register.$global

export const TemplateArgument = Register.global("$hxClasses")["mcl.args.TemplateArgument"] = 
class TemplateArgument extends Register.inherits() {
	new(s, pos) {
		this.expectJsValue = false;
		this.expectBlock = false;
		this.name = s;
		this.pos = pos;
	}
	parseValue(s, p, context) {
		throw Exception.thrown("override this method in subclass, plz thx");
	}
	parseValueBlock(s, p, context) {
		throw Exception.thrown("override this method in subclass, plz thx");
	}
	static parse(s, p) {
		let colon = s.indexOf(":");
		let type = (colon == -1) ? "literal" : s.substring(colon + 1);
		let name = (colon == -1) ? s : s.substring(0, colon);
		if (type == "literal") {
			return new LiteralTemplateArgument(p, name);
		};
		if (!TemplateArgument.argumentTypes.inst.has(type)) {
			throw Exception.thrown("Unknown template argument type: '" + type + "'");
		};
		return Type.createInstance(TemplateArgument.argumentTypes.inst.get(type), [name]);
	}
	static register(type, c) {
		if (TemplateArgument.argumentTypes.inst.has(type)) {
			throw Exception.thrown("Template argument type already registered: " + type);
		};
		TemplateArgument.argumentTypes.inst.set(type, c);
	}
	static get __name__() {
		return "mcl.args.TemplateArgument"
	}
	get __class__() {
		return TemplateArgument
	}
}


TemplateArgument.jsCache = null
TemplateArgument.jsCacheIdx = 0
TemplateArgument.argumentTypes = new StringMap()
