// mc-build/dist/haxpression/CallOperations.js
import {Value} from "./Value.js"
import {ExpressionTypes} from "./ExpressionTypes.js"
import {Error as Error__1} from "./Error.js"
import {StringMap} from "../haxe/ds/StringMap.js"
import {Exception} from "../haxe/Exception.js"
import {Register} from "../genes/Register.js"

const $global = Register.$global

export const CallOperations = Register.global("$hxClasses")["haxpression.CallOperations"] = 
class CallOperations {
	static addFunction(callee, arity, operation) {
		let this1 = CallOperations.map;
		let value = {"arity": arity, "operation": CallOperations.wrapOperation(callee, arity, operation)};
		this1.inst.set(callee, value);
	}
	static hasFunction(callee) {
		return CallOperations.map.inst.has(callee);
	}
	static getFunction(callee) {
		if (!CallOperations.hasFunction(callee)) {
			throw Exception.thrown(new Error__1("no function implementation found for name: " + callee));
		};
		return CallOperations.map.inst.get(callee).operation;
	}
	static getArity(callee) {
		if (!CallOperations.hasFunction(callee)) {
			throw Exception.thrown(new Error__1("no function implementation found for name: " + callee));
		};
		return CallOperations.map.inst.get(callee).arity;
	}
	static canEvaluate(callee, $arguments) {
		if (!CallOperations.hasFunction(callee)) {
			return false;
		};
		let arity = CallOperations.getArity(callee);
		if (arity >= 0 && arity != $arguments.length) {
			return false;
		};
		return ExpressionTypes.canEvaluateAll($arguments);
	}
	static evaluate(callee, $arguments) {
		let operation = CallOperations.getFunction(callee);
		return operation($arguments);
	}
	static wrapOperation(callee, arity, operation) {
		return function ($arguments) {
			if (arity >= 0 && $arguments.length != arity) {
				throw Exception.thrown(new Error__1("function " + callee + " expects exactly " + arity + " argument(s)"));
			};
			return operation($arguments);
		};
	}
	static get __name__() {
		return "haxpression.CallOperations"
	}
	get __class__() {
		return CallOperations
	}
}


;{
	CallOperations.map = new StringMap();
	CallOperations.addFunction("abs", 1, function ($arguments) {
		return Value.fromFloat(Math.abs(Value.toFloat($arguments[0])));
	});
	CallOperations.addFunction("acos", 1, function ($arguments) {
		return Value.fromFloat(Math.acos(Value.toFloat($arguments[0])));
	});
	CallOperations.addFunction("asin", 1, function ($arguments) {
		return Value.fromFloat(Math.asin(Value.toFloat($arguments[0])));
	});
	CallOperations.addFunction("atan", 1, function ($arguments) {
		return Value.fromFloat(Math.atan(Value.toFloat($arguments[0])));
	});
	CallOperations.addFunction("atan2", 2, function ($arguments) {
		return Value.fromFloat(Math.atan2(Value.toFloat($arguments[0]), Value.toFloat($arguments[1])));
	});
	CallOperations.addFunction("ceil", 1, function ($arguments) {
		return Value.fromInt(Math.ceil(Value.toFloat($arguments[0])));
	});
	CallOperations.addFunction("cos", 1, function ($arguments) {
		return Value.fromFloat(Math.cos(Value.toFloat($arguments[0])));
	});
	CallOperations.addFunction("exp", 1, function ($arguments) {
		return Value.fromFloat(Math.exp(Value.toFloat($arguments[0])));
	});
	CallOperations.addFunction("fceil", 1, function ($arguments) {
		return Value.fromFloat(Math.ceil(Value.toFloat($arguments[0])));
	});
	CallOperations.addFunction("ffloor", 1, function ($arguments) {
		return Value.fromFloat(Math.floor(Value.toFloat($arguments[0])));
	});
	CallOperations.addFunction("floor", 1, function ($arguments) {
		return Value.fromInt(Math.floor(Value.toFloat($arguments[0])));
	});
	CallOperations.addFunction("fround", 1, function ($arguments) {
		return Value.fromFloat(Math.round(Value.toFloat($arguments[0])));
	});
	CallOperations.addFunction("log", 1, function ($arguments) {
		return Value.fromFloat(Math.log(Value.toFloat($arguments[0])));
	});
	CallOperations.addFunction("max", 2, function ($arguments) {
		return Value.fromFloat(Math.max(Value.toFloat($arguments[0]), Value.toFloat($arguments[1])));
	});
	CallOperations.addFunction("min", 2, function ($arguments) {
		return Value.fromFloat(Math.min(Value.toFloat($arguments[0]), Value.toFloat($arguments[1])));
	});
	CallOperations.addFunction("pow", 2, function ($arguments) {
		return Value.fromFloat(Math.pow(Value.toFloat($arguments[0]), Value.toFloat($arguments[1])));
	});
	CallOperations.addFunction("random", 0, function ($arguments) {
		return Value.fromFloat(Math.random());
	});
	CallOperations.addFunction("rand", 0, function ($arguments) {
		return Value.fromFloat(Math.random());
	});
	CallOperations.addFunction("round", 1, function ($arguments) {
		return Value.fromInt(Math.round(Value.toFloat($arguments[0])));
	});
	CallOperations.addFunction("sin", 1, function ($arguments) {
		return Value.fromFloat(Math.sin(Value.toFloat($arguments[0])));
	});
	CallOperations.addFunction("sqrt", 1, function ($arguments) {
		return Value.fromFloat(Math.sqrt(Value.toFloat($arguments[0])));
	});
	CallOperations.addFunction("tan", 1, function ($arguments) {
		return Value.fromFloat(Math.tan(Value.toFloat($arguments[0])));
	});
}
