// mc-build/dist/haxpression/ValueType.js
import {Register} from "../genes/Register.js"

const $global = Register.$global

export const ValueType = 
Register.global("$hxEnums")["haxpression.ValueType"] = 
{
	__ename__: "haxpression.ValueType",
	
	VFloat: Object.assign((v) => ({_hx_index: 0, __enum__: "haxpression.ValueType", "v": v, "@kind": "VFloat"}), {_hx_name: "VFloat", __params__: ["v"]}),
	VInt: Object.assign((v) => ({_hx_index: 1, __enum__: "haxpression.ValueType", "v": v, "@kind": "VInt"}), {_hx_name: "VInt", __params__: ["v"]}),
	VBool: Object.assign((v) => ({_hx_index: 2, __enum__: "haxpression.ValueType", "v": v, "@kind": "VBool"}), {_hx_name: "VBool", __params__: ["v"]}),
	VString: Object.assign((v) => ({_hx_index: 3, __enum__: "haxpression.ValueType", "v": v, "@kind": "VString"}), {_hx_name: "VString", __params__: ["v"]}),
	VNA: {_hx_name: "VNA", _hx_index: 4, __enum__: "haxpression.ValueType", "@kind": "VNA"},
	VNM: {_hx_name: "VNM", _hx_index: 5, __enum__: "haxpression.ValueType", "@kind": "VNM"}
}
ValueType.__constructs__ = [ValueType.VFloat, ValueType.VInt, ValueType.VBool, ValueType.VString, ValueType.VNA, ValueType.VNM]
ValueType.__empty_constructs__ = [ValueType.VNA, ValueType.VNM]
