// mc-build/dist/haxpression/Value.js
import {ValueType} from "./ValueType.js"
import {Error as Error__1} from "./Error.js"
import {Exception} from "../haxe/Exception.js"
import {Register} from "../genes/Register.js"

const $global = Register.$global

export const Value = Register.global("$hxClasses")["haxpression._Value.Value"] = 
class Value {
	static fromValueType(valueType) {
		return valueType;
	}
	static toValueType(this1) {
		return this1;
	}
	static fromFloat(v) {
		return Value.fromValueType(ValueType.VFloat(v));
	}
	static fromInt(v) {
		return Value.fromValueType(ValueType.VInt(v));
	}
	static fromBool(v) {
		return Value.fromValueType(ValueType.VBool(v));
	}
	static toFloat(this1) {
		switch (this1._hx_index) {
			case 0:
				let v = this1.v;
				return v;
				break
			case 1:
				let v1 = this1.v;
				return v1;
				break
			case 2:
				let v2 = this1.v;
				throw Exception.thrown(new Error__1("cannot convert Bool to Float"));
				break
			case 3:
				let v3 = this1.v;
				throw Exception.thrown(new Error__1("cannot convert String to Float"));
				break
			case 4:
				throw Exception.thrown(new Error__1("cannot convert NA to Float"));
				break
			case 5:
				throw Exception.thrown(new Error__1("cannot convert NM to Float"));
				break
			
		};
	}
	static toInt(this1) {
		switch (this1._hx_index) {
			case 0:
				let v = this1.v;
				return v | 0;
				break
			case 1:
				let v1 = this1.v;
				return v1;
				break
			case 2:
				let v2 = this1.v;
				throw Exception.thrown(new Error__1("cannot convert Bool to Int"));
				break
			case 3:
				let v3 = this1.v;
				throw Exception.thrown(new Error__1("cannot convert String to Int"));
				break
			case 4:
				throw Exception.thrown(new Error__1("cannot convert NA to Int"));
				break
			case 5:
				throw Exception.thrown(new Error__1("cannot convert NM to Int"));
				break
			
		};
	}
	static toBool(this1) {
		switch (this1._hx_index) {
			case 0:
				let v = this1.v;
				return v != 0.0;
				break
			case 1:
				let v1 = this1.v;
				return v1 != 0;
				break
			case 2:
				let v2 = this1.v;
				return v2;
				break
			case 3:
				let v3 = this1.v;
				return v3.toLowerCase() == Value.TRUE_STRING;
				break
			case 4:
				throw Exception.thrown(new Error__1("cannot convert NA to Bool"));
				break
			case 5:
				throw Exception.thrown(new Error__1("cannot convert NM to Bool"));
				break
			
		};
	}
	static isNA(this1) {
		if (this1._hx_index == 4) {
			return true;
		} else {
			return false;
		};
	}
	static isNM(this1) {
		if (this1._hx_index == 5) {
			return true;
		} else {
			return false;
		};
	}
	static get __name__() {
		return "haxpression._Value.Value_Impl_"
	}
	get __class__() {
		return Value
	}
}


Value.TRUE_STRING = "true"
